### PR TITLE
GlobalEventsHandlerProvider: new provider to share global event handlers to consuming components

### PR DESCRIPTION
**Pull request by AlbertCarreras**
#### What changed?

GlobalEventsHandlerProvider: new provider to share global event handlers to consuming components #2959

Brave Browser - HandlerProvider - Gestalt 2023-05-23 at 10 52 54 AM

#### Why?

Read RFC above 

### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-XXXX)
- [TDD](link to Paper doc)
- [Figma](link to Figma file)

### Checklist

- [ ] Added unit and Flow Tests
- [ ] Added documentation + accessibility tests
- [ ] Verified accessibility: keyboard & screen reader interaction
- [ ] Checked dark mode, responsiveness, and right-to-left support
- [ ] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

**Comment by rlingineni** [comment]
Hi, this is great!

It seems like it's up to Pinboard subscribers to individually define each element. I wasn't sure if the plan is to have this be globally added to Pinboard at the top-level, or re-used across different places. Are things getting overrided a possible concern? 

Like is this a valid case?
```
// app.js
<MyApp>
<GlobalEventsHandlerProvider button={} sheetMobile={// maybe not called } >
    <Button>
    <GlobalEventsHandlerProvider sheetMobile={} >
      <SheetMobile>
      <Button>
  <GlobalEventsHandlerProvider>
<GlobalEventsHandlerProvider>
<MyApp/>
```

It does seem like for every component, we'd have to go in and add it as a potential [prop](https://deploy-preview-2959--gestalt.netlify.app/web/utilities/globaleventshandlerprovider#Props) and keep it up to date with our events getting changed. 


What if, and not completely sure of the exact trade offs, but if the event emitting was more generic like an `onEvent` handler and not tuned by component? The caller can then handle it based on their needs. In the future, we could also expand to emit non interactive events (e.g time to paint, or mount) as part of that emitter. 

```
<MyApp>
<GlobalEventsHandlerProvider onEvent={} >
    <Button>
    <GlobalEventsHandlerProvider onEvent={} >
      <SheetMobile>
      <Button>
  <GlobalEventsHandlerProvider>
<GlobalEventsHandlerProvider>
<MyApp/>
```

And on Event could give everything like:
```js
const onEvent = (event:string, sender:name, message?:any)  => {
 console.log(`A gestalt event happened for ${sender}`)
}
```
And in all the components we call something like

```
emit('onClick', 'TextField', {value:'H'})
emit('onDismiss', 'SheetMobile', {value:'H'})
```

File: docs/examples/globaleventshandlerprovider/sheetMobileHandlers.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+export default function Example(): Node {
+  const [showComponent, setShowComponent] = useState<null | boolean>(true);
+
+  const PAGE_HEADER_ZINDEX: FixedZIndex = new FixedZIndex(10);
+  const ABOVE_PAGE_HEADER_ZINDEX: CompositeZIndex = new CompositeZIndex([PAGE_HEADER_ZINDEX]);
+
+  return (
+    <GlobalEventsHandlerProvider
+      sheetMobile={{
```
Comment (dangerismycat) [inline docs/examples/globaleventshandlerprovider/sheetMobileHandlers.js:23]: Should we demonstrate proper Provider value best practices in the examples, by memoizing the passed-down objects and contained callbacks?
```js
const sheetMobileOnOpen = useCallback(() => console.log('On open handler'), []);
const sheetMobileOnClose = useCallback(() => console.log('On close handler'), []);
const sheetMobileHandlers = useMemo(() => ({ onOpen: sheetMobileOnOpen, onClose: sheetMobileOnClose }), []);

...

<GlobalEventsHandlerProvider sheetMobile={sheetMobileHandlers}>
```
const sheetMobileHandlers =

File: docs/pages/web/sheetmobile.js [hunk 2]
```diff
@@ -327,6 +328,27 @@ When using these render props, just pass the argument \`onDismissStart\` to your
             }
           />
         </MainSection.Subsection>
+
+        <MainSection.Subsection
+          title="External handlers"
+          description={`SheetMobile consumes external handlers from GlobalEventsHandlerProvider.
+
+Handlers:
+- onOpen: executed on useEffect
```
Comment (dangerismycat) [inline docs/pages/web/sheetmobile.js:337]: "executed on useEffect" doesn't really mean anything. An effect will get run once when the component mounts, but then also again any time its dependencies change. I wouldn't get caught up on the use of `useEffect` here — that's really an implementation detail. What's pertinent to the dev is that `onOpen` will run when the sheet opens, and that `onClose` will run just before the sheet closes. Whether it uses `useEffect` or some other mechanism is irrelevant — and really, if the developer needs to be be aware of that mechanism, we have a leaky abstraction.

File: docs/pages/web/sheetmobile.js [hunk 3]
```diff
@@ -327,6 +328,27 @@ When using these render props, just pass the argument \`onDismissStart\` to your
...
+- onClose: executed on the cleanup function of useEffect
+
+See [GlobalEventsHandlerProvider](/web/utilities/GlobalEventsHandlerProvider) for more information.`}
```
Comment (dangerismycat) [inline docs/pages/web/sheetmobile.js:340]: This URL isn't correct — component names in URLs are always all lowercase

File: packages/gestalt/src/SheetMobile/FullPage.js [hunk 4]
```diff
@@ -61,8 +62,25 @@ export default function FullPage({
...
+  // Consumes DefaultLabelProvider
   const { accessibilityLabel: defaultAccessibilityLabel } = useDefaultLabelContext('SheetMobile');
 
+  // Consumes GlobalEventsHandlerProvider
+  const { sheetMobile } = useGlobalEventsHandlerContext() ?? {
+    sheetMobile: { onOpen: () => {}, onClose: () => {} },
+  };
+
+  useEffect(() => {
+    sheetMobile?.onOpen?.();
```
Comment (dangerismycat) [inline packages/gestalt/src/SheetMobile/FullPage.js:76]: This reads a little wonky. Consider renaming the variable from useGlobalEventsHandlerContext to something like "sheetMobileHandlers" so this reads more like `sheetMobileHandlers?.onOpen?.()`

File: packages/gestalt/src/SheetMobile/FullPage.js [hunk 5]
```diff
@@ -61,8 +62,25 @@ export default function FullPage({
...
+
+    return function cleanup() {
+      sheetMobile?.onClose?.();
+    };
+  }, [sheetMobile]);
```
Comment (dangerismycat) [inline packages/gestalt/src/SheetMobile/FullPage.js:81]: Passing an object to `useEffect` isn't great, as the object reference can stay the same while its properties change, and the effect won't run again. We should destructure the handlers out of `sheetMobileHandlers` and use those directly in the dependency array.

File: packages/gestalt/src/SheetMobile/PartialPage.js [hunk 6]
```diff
@@ -85,10 +86,25 @@ export default function PartialPage({
...
+    sheetMobile: { onOpen: () => {}, onClose: () => {} },
+  };
+
+  useEffect(() => {
+    sheetMobile?.onOpen?.();
+
+    return function cleanup() {
+      sheetMobile?.onClose?.();
+    };
+  }, [sheetMobile]);
```
Comment (dangerismycat) [inline packages/gestalt/src/SheetMobile/PartialPage.js:105]: ditto here on renaming and destructuring

File: packages/gestalt/src/contexts/GlobalEventsHandlerProvider.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+// @flow strict
+import { createContext, type Context, type Element, type Node, useContext } from 'react';
+
+export type NoopType = () => void;
+
+type SheetMobileHandlersType = {|
```
Comment (dangerismycat) [inline packages/gestalt/src/contexts/GlobalEventsHandlerProvider.js:6]: nit: we're not exporting this type or using it elsewhere in this file, so I'd just inline it in GlobalEventsHandlerContextType

File: playwright/accessibility/web__utilities__globaleventshandlerprovider.spec.mjs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// @flow strict
+import { test } from '@playwright/test';
+import expectAccessiblePage from './expectAccessiblePage.mjs';
+
+test('GlobalEventsHandlerProvider Accessibility check', async ({ page }) => {
+  await page.goto('/web/utilities/GlobalEventsHandlerProvider');
```
Comment (dangerismycat) [inline playwright/accessibility/web__utilities__globaleventshandlerprovider.spec.mjs:6]: This URL isn't correct

File: docs/pages/web/utilities/globaleventshandlerprovider.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        type="utility"
+      />
+
+      <GeneratedPropTable generatedDocGen={generatedDocGen} />
+
+      <MainSection name="Variants">
+        <MainSection.Subsection title="GlobalEventsHandlerProvider in SheetMobile">
+          <MainSection.Card
+            cardSize="lg"
+            description={`GlobalEventsHandlerProvider has one prop for each component subscribing to the provider.
```
Comment (dangerismycat) [inline docs/pages/web/utilities/globaleventshandlerprovider.js:26]: This copy needs to be updated. It still references the previous Provider name, and is a bit vague about if this affects all usages of a single component or if it needs to be implemented at the component callsite. Consider something like
"GlobalEventsHandlerProvider is a React Context Provider that is intended to be implemented at the top level of your app or feature. Each prop corresponds to a specific Gestalt component, and allows you to add logic that will run for every instance of that component in the component tree.

For SheetMobile, we can add `onOpen` and `onClose` handlers. `onOpen` runs on component mount, when the SheetMobile first appears in the UI. `onClose` runs just before the component unmounts and is removed from the UI."

**Review by dangerismycat — APPROVED** [review]

File: docs/docs-components/COMPONENT_DATA.js [hunk 10]
```diff
@@ -3976,6 +3977,59 @@ const UTILITIES_LIST: $ReadOnlyArray<ListItemType> = [
       figma: null,
     },
   },
+  {
+    svg: <ProviderHandlers />,
+    name: 'GlobalEventsHandlerProvider',
+    aliases: ['React.useEffect'],
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply docs/docs-components/COMPONENT_DATA.js:3983]: NO, good ctch

File: rfcs/rfcs/2023-05-effectsProvider-new-component/README.md [hunk 11]
```diff
@@ -0,0 +1,335 @@
+- Start Date: May 2023
+- RFC PR: https://github.com/pinterest/gestalt/pull/2918
+- Authors: Alberto Carreras, acarreras
+
+# (RFC title goes here)
+
+## Summary
+
+EffectsProvider is a React.Context provider intended to share React effects (useEffect logic) with consuming components.
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply rfcs/rfcs/2023-05-effectsProvider-new-component/README.md:9]: yes

File: docs/docs-components/COMPONENT_DATA.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply docs/docs-components/COMPONENT_DATA.js:3983]: DONE

File: docs/pages/web/sheetmobile.js (hunk 3 above)
Comment (AlbertCarreras) [reply docs/pages/web/sheetmobile.js:340]: DONE

File: docs/pages/web/utilities/globaleventshandlerprovider.js (hunk 9 above)
Comment (AlbertCarreras) [reply docs/pages/web/utilities/globaleventshandlerprovider.js:26]: DONE

File: packages/gestalt/src/SheetMobile/FullPage.js (hunk 4 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/SheetMobile/FullPage.js:76]: DONE

File: packages/gestalt/src/SheetMobile/FullPage.js (hunk 5 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/SheetMobile/FullPage.js:81]: DONE

**Pull request merged**